allennlp/models/reading_comprehension/dialog_qa.py

```diff
@@ -0,0 +1,407 @@
+        self_atten_vecs = torch.matmul(self_atten_probs, residual_layer)
+
+        residual_layer = F.relu(self._merge_self_atten(torch.cat([self_atten_vecs, residual_layer,
+                                                                  residual_layer * self_atten_vecs],
+                                                                 dim=-1)))
+
+        final_merged_passage = final_merged_passage + residual_layer  # batch_size * maxqa_pair_len * max_passage_len * 200
+        final_merged_passage = self._variational_dropout(final_merged_passage)
+        start_rep = self._span_start_encoder(final_merged_passage, passage_mask)
+        span_start_logits = self._span_start_predictor(start_rep).squeeze(-1)
+
+        end_rep = self._span_end_encoder(torch.cat([final_merged_passage, start_rep], dim=-1), passage_mask)
+        span_end_logits = self._span_end_predictor(end_rep).squeeze(-1)
+
+        span_yesno_logits = self._span_yesno_predictor(end_rep).squeeze(-1)
+        span_followup_logits = self._span_followup_predictor(end_rep).squeeze(-1)
+
+        span_start_logits = util.replace_masked_values(span_start_logits, passage_mask, -1e7)
+        # batch_size * maxqa_len_pair, max_document_len
+        span_end_logits = util.replace_masked_values(span_end_logits, passage_mask, -1e7)
+
+        best_span = self._get_best_span_yesno_followup(span_start_logits, span_end_logits,
+                                                       span_yesno_logits, span_followup_logits)
+
+        output_dict: Dict[str, Any] = {}
+
+        # Compute the loss.
+        if span_start is not None:
+            loss = nll_loss(util.masked_log_softmax(span_start_logits, passage_mask), span_start.view(-1),
+                            ignore_index=-1)
+            self._span_start_accuracy(span_start_logits, span_start.view(-1), mask=qa_mask)
+            loss += nll_loss(util.masked_log_softmax(span_end_logits,
+                                                     passage_mask), span_end.view(-1), ignore_index=-1)
+            self._span_end_accuracy(span_end_logits, span_end.view(-1), mask=qa_mask)
+            self._span_accuracy(best_span[:, 0:2],
+                                torch.stack([span_start, span_end], -1).view(batch_size * max_qa_count, 2),
+                                mask=qa_mask.unsqueeze(1).expand(-1, 2).long())
+            # add a select for the right span to compute loss
+            gold_span_end_loc = []
+            span_end = span_end.view(-1).squeeze().data.cpu().numpy()
+            for i in range(0, batch_size * max_qa_count):
+                gold_span_end_loc.append(max(span_end[i] * 3 + i * passage_length * 3, 0))
+                gold_span_end_loc.append(max(span_end[i] * 3 + i * passage_length * 3 + 1, 0))
+                gold_span_end_loc.append(max(span_end[i] * 3 + i * passage_length * 3 + 2, 0))
+            gold_span_end_loc = torch.LongTensor(gold_span_end_loc)
```

Comment:
Instead of this, you can do `gold_span_end_loc = span_end.new(gold_span_end_loc)`, which will create the tensor on the right device. Same with the predicted tensor below.